### PR TITLE
VideoCommon: allow multiple texture assets to associate with a texture cache entry

### PR DESCRIPTION
**Pull request by iwubcode**
(used in a future review)

### CODE REVIEW

**Comment by AdmiralCurtiss** [comment]
Can you explain why this is useful and what you plan to use it for?

**Comment by iwubcode** [comment]
@AdmiralCurtiss - apologies, I should have explained it in the opening.  I was being lazy.  I'm going to give you a bit of TMI but I want to explain things so that you have a better understanding of where I'm going.

First, I'll distinguish between a "texture" (the thing the graphics API loads onto the GPU) and an image, what you might author in Photoshop.

Images can be added as a singular texture, much like Dolphin does for its game textures that exist in the Texture Cache or you can add a separate image to an existing texture as another _layer_.  Dolphin has support for layers because it uses them for stereo 3d to handle multiple views.  A layer has the requirement that the image must be the same size and the same format as well.

In addition to a texture, there's a thing called a _sampler_.  The sampler tells a shader how to sample the texture!

When you provide a shader details about a texture, you also provide a sampler.  Samplers are limited, even on modern GPUs you can only have something like 16.  Technically you can have many many textures which is why we can have so many in the TextureCache!

Ok, so at the moment Dolphin supports a single custom texture to be loaded.  What this does, is instead of loading the game texture into layer 0, we load the custom texture instead!

But what if we want more images?  Modern lighting implementations can use a bunch of images!

We could load multiple images, each as its own texture.  But we'd have to expose it to the shader with individual samplers.  Dolphin already uses up to 8 samplers on Wii/GC for a single draw.  If we wanted to do something _crazy_ we'd either need more (remember we can only have at most 16) or we'd have to know how many Dolphin was using and use whatever was left.  I do want to support more samplers in the [near future](https://github.com/dolphin-emu/dolphin/pull/11783) but not _now_.

The other option, is to load the other images as separate layers in the existing game texture!  That is what is happening in this PR.

The other thing this PR does is associate this texture cache entry with not just the game texture asset but also all these additional images as assets as well.  This means if I go and tweak my Nth image, it will reload in Dolphin automatically.

This PR is ultimately to support an updated #11300 .  In the current implementation of that PR, I append the additional layers _after_ Dolphin loads the custom texture.  This is very inefficient because to create a texture, you have to tell it how many layers you intend to have up front.

Please let me know if you have any questions.  I'm sorry for the wall of text.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 1]
```diff
@@ -1633,20 +1644,34 @@ RcTcacheEntry TextureCacheBase::CreateTextureEntry(
...
 
   RcTcacheEntry entry;
-  if (custom_texture_data && custom_texture_data->m_levels.size() >= 1)
-  {
-    const u32 texLevels = no_mips ? 1 : (u32)custom_texture_data->m_levels.size();
-    const auto& first_level = custom_texture_data->m_levels[0];
-    const TextureConfig config(first_level.width, first_level.height, texLevels, 1, 1,
-                               first_level.format, 0);
+  if (!assets_data.empty() &&
+      std::none_of(assets_data.begin(), assets_data.end(), [](auto asset) { return !asset; }))
```
Comment (AdmiralCurtiss) [inline Source/Core/VideoCommon/TextureCacheBase.cpp:1648]: `(const auto& asset)`. I'm also not sure if this condition is needed at all, this should never get empty shared_ptrs in assets_data, no?

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 2]
```diff
@@ -1633,20 +1644,34 @@ RcTcacheEntry TextureCacheBase::CreateTextureEntry(
...
+  {
+    const auto calculate_max_levels = [&]() {
+      const auto max_element =
+          std::max_element(assets_data.begin(), assets_data.end(), [](auto lhs, auto rhs) {
```
Comment (AdmiralCurtiss) [inline Source/Core/VideoCommon/TextureCacheBase.cpp:1652]: Here too, `const auto& lhs, const auto& rhs`

File: Source/Core/VideoCommon/TextureCacheBase.cpp (hunk 1 above)
Comment (AdmiralCurtiss) [reply Source/Core/VideoCommon/TextureCacheBase.cpp:1648]: If you do want to check for this put it in an assert so we know this is a logic error.

**Pull request merged**